### PR TITLE
Adding test cases for null point classification

### PR DESCRIPTION
**Pull request by haman80**
- [x] I have added a changelog entry for this pull request.

- [x] If adding new functionality, I have added tests and
      docstrings.

- [x] I have fixed any newly failing tests.

### CODE REVIEW

File: docs/notebooks/analysis/nullpoint.ipynb [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    "On the other hand, `null_point_find()` is used when the user does not have an adequate modeling function. It also does not require the grid to have uniform spacing in each dimension. Instead, it will ask the user for three arrays (one for each dimension) of coordinates that determines the desired custom spacing in each of the three dimensions, and then constructs the resulting grid on its own. Furthermore, it requires the user to input all of the three components of magnetic field values, each as a 3D array with the same size as the grid, with each entry representing the strength of that component of the magnetic field for the corresponding grid coordinate. \n",
+    "Finally, both functions take in as arguments two convergence thresholds that we will discuss later, for locating the null points. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "72333e74-442c-4f7a-a3c9-76fad88eaac2",
+   "metadata": {},
+   "source": [
+    "## Locating a null point\n",
```
Comment (namurphy) [inline docs/notebooks/analysis/nullpoint.ipynb:66]:
```suggestion
    "### Locating a null point\n",
```

File: docs/notebooks/analysis/nullpoint.ipynb [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    "2) Trilinear Analysis: Assuming a trilinear field, the cells that have passed the reduction check are then analyzed, so that that we can be sure if they do contain a null point. \n",
+    "3) Locating the null point: The cell that contains a null point is isolated, and the location of the null point is estimated using the iterative Newton-Raphson method with an initial random guess."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2bd7ab2d-427f-4d85-9d58-352aeffc0b52",
+   "metadata": {},
+   "source": [
+    "## Classifying a null point\n",
```
Comment (namurphy) [inline docs/notebooks/analysis/nullpoint.ipynb:78]:
```suggestion
    "### Classifying a null point\n",
```

File: docs/notebooks/analysis/nullpoint.ipynb [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    "Classification is done by analysing the jacobian matrix calculated at the location of null point. The full method is explained in the paper by [Parnell et al. (1996)](https://aip.scitation.org/doi/10.1063/1.871810)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "bdc1d3a6-35f3-4b6c-ac59-f81596db18ab",
+   "metadata": {},
+   "source": [
+    "# Running through examples\n",
```
Comment (namurphy) [inline docs/notebooks/analysis/nullpoint.ipynb:87]:
```suggestion
    "## Running through examples\n",
```

File: docs/notebooks/analysis/nullpoint.ipynb [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    "We will now run through an example for each of the two null point finding functions to see how to properly utilize the null point finder. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ebcced41-d08a-405e-9f7e-df89369bca9f",
+   "metadata": {},
+   "source": [
+    "# Uniform regular grid with a model function\n",
```
Comment (namurphy) [inline docs/notebooks/analysis/nullpoint.ipynb:96]:
```suggestion
    "### Uniform regular grid with a model function\n",
```

File: docs/notebooks/analysis/nullpoint.ipynb [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+   "source": [
+    "As we can see `uniform_null_point_find()` correctly identifies the location of the null point in addition to its type, which is a proper radial null.  "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5b30dd51-d457-4a4b-b9d8-0471f13abb7e",
+   "metadata": {},
+   "source": [
+    "# Arbitary Regular Grid\n",
```
Comment (namurphy) [inline docs/notebooks/analysis/nullpoint.ipynb:162]:
```suggestion
    "### Arbitary Regular Grid\n",
```

File: docs/notebooks/analysis/nullpoint.ipynb [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    "print(npoints2[0].loc)\n",
+    "print(npoints2[0].type)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "95cfcba7-2531-463b-88f5-40b60056602f",
+   "metadata": {},
+   "source": [
+    "As we can see the magentic field provided above has a Spiral null point located at $(5.5,5.5,5.5)$"
```
Comment (namurphy) [inline docs/notebooks/analysis/nullpoint.ipynb:202]:
```suggestion
    "As we can see the magnetic field provided above has a spiral null point located at $(5.5,5.5,5.5)$"
```

**Comment by haman80** [comment]
pre-commit.ci autofix

File: docs/notebooks/analysis/nullpoint.ipynb [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+   "metadata": {},
+   "source": [
+    "## Contents:\n",
+    "\n",
+    "1. [How Null Point Finder works](#How-Null-Point-Finder-works)\n",
+    "    1. [Locating a null point](#Locating-a-null-point)\n",
+    "    1. [Classifying a null point](#Classifying-a-null-point)\n",
+    "1. [Running through examples](#Running_through-examples)\n",
+    "    1. [Uniform regular grid with a model function](#Uniform-regular-grid-with-a-model-function)\n",
+    "    1. [Arbitary Regular Grid](#Arbitary-regular-grid)"
```
Comment (namurphy) [inline docs/notebooks/analysis/nullpoint.ipynb:47]:
```suggestion
    "    1. [Arbitrary Regular Grid](#Arbitrary-regular-grid)"
```

File: docs/notebooks/analysis/nullpoint.ipynb [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+   "source": [
+    "As we can see `uniform_null_point_find()` correctly identifies the location of the null point in addition to its type, which is a proper radial null.  "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5b30dd51-d457-4a4b-b9d8-0471f13abb7e",
+   "metadata": {},
+   "source": [
+    "### Arbitary Regular Grid\n",
```
Comment (namurphy) [inline docs/notebooks/analysis/nullpoint.ipynb:162]:
```suggestion
    "### Arbitrary Regular Grid\n",
```

File: changelog/1496.feature.rst [hunk 9]
```diff
@@ -0,0 +1,2 @@
+Added a null point classifier function which determines the
+type of a given 3D magnetic null point.
```
Comment (namurphy) [inline changelog/1496.feature.rst:2]: ...update the number in the file to the most recent PR?

Comment (namurphy) [reply changelog/1496.feature.rst:2]: And add `1554.doc.rst` to say that you added an example notebook?

**Review by namurphy — COMMENTED** [review]
It's looking really close!  The main things are...
 - Updating the changelog entries
 - Deciding if the tolerances in a few places could be lowered from `1e-4` to perhaps `1e-10`, given the fixes that you've made recently
 - Possibly adding some assertions for the last few tests, for example about the null point classification.

Apart from that, most of my suggestions are minor (i.e. removing commented out code).  Thank you again, and hope you have a good weekend!

**Comment by haman80** [comment]
pre-commit.ci autofix

**Comment by haman80** [comment]
pre-commit.ci autofix

File: plasmapy/analysis/tests/test_nullpoint.py [hunk 10]
```diff
@@ -346,23 +349,133 @@ def test_null_point_find7():
...
+def test_null_point_find11():
+    nullpoint10_args = {
+        "x_range": [-0.1, 0.1],
+        "y_range": [-0.1, 0.1],
+        "z_range": [-0.1, 0.1],
+        "precision": [0.01, 0.01, 0.01],
+        "func": lambda x, y, z: [1.01 * y * z, -x * z, x * y],
+    }
+    npoints = uniform_null_point_find(**nullpoint10_args)
+    assert True
```
[the remark this answers is not in the document]
Comment (haman80) [reply plasmapy/analysis/tests/test_nullpoint.py:480]: I am not exactly sure how we would go about doing so, considering that there are many null points found. 
I could add assertions that check the classification of a couple of manually selected null points I suppose, but I am not exactly sure if that is the best way to go about it.

File: plasmapy/analysis/tests/test_nullpoint.py [hunk 11]
```diff
@@ -318,8 +321,8 @@ def test_null_point_find5():
         "precision": [0.01, 0.01, 0.01],
         "func": vspace_func_4,
     }
-    npoints5 = uniform_nullpoint_find(**nullpoint5_args)
-    assert len(npoints5) == 0
+    npoints5 = uniform_null_point_find(**nullpoint5_args)
+    assert True
```
[the remark this answers is not in the document]
Comment (haman80) [reply plasmapy/analysis/tests/test_nullpoint.py:325]: Similarly to the other comment, I am not exactly sure how we would go about doing so, considering that there are many null points found.

**Review by namurphy — APPROVED** [review]
